python/ray/train/v2/_internal/execution/controller/placement_group_cleaner.py

```diff
@@ -0,0 +1,149 @@
+import logging
+import threading
+import time
+from typing import Optional
+
+import ray
+from ray.train.v2._internal.constants import GET_ACTOR_TIMEOUT_S
+from ray.train.v2._internal.state.util import is_actor_alive
+from ray.util.placement_group import PlacementGroup, remove_placement_group
+
+logger = logging.getLogger(__name__)
+
+
+class PlacementGroupCleaner:
+    """Detached helper that ensures PG cleanup if Ray Train Controller exits ungracefully.
+
+    This actor should be created with lifetime='detached' to avoid being
+    fate-shared with the Train controller.
+    """
+
+    def __init__(self, check_interval_s: float = 1.0):
+        self._check_interval_s = check_interval_s
+        self._stopped: bool = False
+        self._controller_actor_id: Optional[str] = None
+        self._placement_group: Optional[PlacementGroup] = None
+        self._monitoring: bool = False
+        self._monitor_thread: Optional[threading.Thread] = None
+        self._get_actor_timeout_s = GET_ACTOR_TIMEOUT_S
+        self._exiting: bool = False
+
+    def register_controller_and_placement_group(
+        self, controller_actor_id: str, placement_group: PlacementGroup
+    ):
+        self._controller_actor_id = controller_actor_id
+        self._placement_group = placement_group
+        logger.info(
+            "PlacementGroupCleaner registered controller %s with placement group %s",
+            controller_actor_id,
+            placement_group.id,
+        )
+
+    def start_monitoring(self):
+        """Start monitoring the controller and placement group."""
+        if not self._controller_actor_id or not self._placement_group:
+            logger.warning(
+                "Cannot start monitoring: controller or placement group missing"
+            )
+            return False
+
+        if self._monitoring:
+            logger.warning("Already monitoring")
+            return False
+
+        if self._stopped:
+            logger.warning("Cannot start monitoring: stop already requested")
+            return False
+
+        self._monitoring = True
+        self._monitor_thread = threading.Thread(
+            target=self._monitor_loop,
+            name="PlacementGroupCleanerMonitor",
+            daemon=True,
+        )
+        self._monitor_thread.start()
+        logger.info("PlacementGroupCleaner started monitoring in background thread")
+        return True
+
+    def _monitor_loop(self):
+        """Monitor controller; remove PG when controller is gone.
+
+        This runs continuously until controller dies or stop() is called.
+        """
+        while not self._stopped and self._controller_actor_id:
+            try:
+                alive = is_actor_alive(
+                    actor_id=self._controller_actor_id,
+                    timeout=self._get_actor_timeout_s,
+                )
```

Comment:
This state api call can also raise. See here for an example error handling of state api: https://github.com/ray-project/ray/blob/d972b7d84cab6d0391e6fd9c107c7f32c1881231/python/ray/train/v2/_internal/state/state_actor.py#L107C31-L107C35

Reply:
also, the current try catch is capturing too much. Let's move the safe logger/sleep calls out of the try catch and only try/catch the state api calls individually